lib/kdl/include/kdl/expected.h

```diff
@@ -0,0 +1,627 @@
+ documentation files (the "Software"), to deal in the Software without restriction, including without limitation the
+ rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit
+ persons to whom the Software is furnished to do so, subject to the following conditions:
+
+ The above copyright notice and this permission notice shall be included in all copies or substantial portions of the
+ Software.
+
+ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
+ WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+ OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
+
+#ifndef result_h
+#define result_h
+
+#include "kdl/overload.h"
+#include "kdl/expected_forward.h"
+
+#include <absl/types/optional.h>
+#include <absl/types/variant.h>
+
+namespace kdl {
+    /**
+     * Wrapper class that can contain either a value or one of several errors.
+     *
+     * An instance of this class represents an expectation for the result of applying a function if that function
+     * returns either a value or throws an error.
+     *
+     * An expectation is considered successful if it contains a value, and a failure if it contains an error.
+     *
+     * @tparam Value the type of the value
+     * @tparam Errors the types of the possible errors
+     */
+    template <typename Value, typename... Errors>
+    class [[nodiscard]] expected {
+    public:
+        using value_type = Value;
+    private:
+        absl::variant<value_type, Errors...> m_value;
+    public:
+        /**
+         * Creates an expectation that wraps the given value, which can be a value or an error.
+         *
+         * If the value or error is passed by rvalue reference, it is moved into this expectation.
+         *
+         * @tparam T the type of the given value or error
+         * @param t the value or error to wrap
+         */
+        template <typename T>
+        expected(T&& t)
+        : m_value(std::forward<T>(t)) {}
+        
+        /**
+         * Applies the given visitor to the given expectation and returns the result returned by the visitor.
+         * The value or error contained in the given expectation is passed to the visitor by const lvalue reference.
+         *
+         * @tparam Visitor the type of the visitor
+         * @param visitor the visitor to apply
+         * @param expected the expectation to apply the visitor to
+         * @return the result of applying the given visitor
+         */
+        template <typename Visitor>
+        friend auto visit(Visitor&& visitor, const expected& expected) {
```

Comment:
Is it just a matter of style that this is a free function?
If it is, I prefer member functions - it gives you more information when reading calling code, but I'm ok with either.

Reply:
A matter of style and consistency with `std::visit`. I'd be ok with a member function, but that deviates from the style of the STL. Personally, I also tend to use more free functions now.

Reply:
Having thought about it more, I agree now since it is more consistent with `map` being a member function as well. Maybe we should rename the function, too, to distinguish it from `std::visit`. Any ideas about an alternative name?

Reply:
Not sure. I think a member function `result.visit(..)` would be OK, it's not a error to reuse the name is it?

Reply:
One reason for a free function is that it's syntactically nicer to operator on an rvalue this way. Consider:

```
auto result = ...;
visit(overload{...}, std::move(result));
```

vs.

```
auto result = ...;
std::move(result).visit(overload{...});
```

I'll keep it a member function for now, but we might want to revert it if this becomes awkward to use.